File: skills/wechatpay-basic-payment/references/1-商户/示例代码/Java/8-订单退款/CreateAbnormalRefund.java
package com.java.demo;

import com.java.utils.WXPayUtility; // 引用微信支付工具库，参考：https://pay.weixin.qq.com/doc/v3/merchant/4014931831

import com.google.gson.annotations.SerializedName;
import com.google.gson.annotations.Expose;
import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;

import java.io.IOException;
import java.io.UncheckedIOException;
import java.security.PrivateKey;
import java.security.PublicKey;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * 发起异常退款
 */
public class CreateAbnormalRefund {
  private static String HOST = "https://api.mch.weixin.qq.com";
  private static String METHOD = "POST";
  private static String PATH = "/v3/refund/domestic/refunds/{refund_id}/apply-abnormal-refund";

  public static void main(String[] args) {
    // TODO: 请准备商户开发必要参数，参考：https://pay.weixin.qq.com/doc/v3/merchant/4013070756
    CreateAbnormalRefund client = new CreateAbnormalRefund(
      "YOUR_MCHID",                    // 商户号，是由微信支付系统生成并分配给每个商户的唯一标识符，商户号获取方式参考 https://pay.weixin.qq.com/doc/v3/merchant/4013070756
      "YOUR_CERT_SERIAL_NO",         // 商户API证书序列号，如何获取请参考 https://pay.weixin.qq.com/doc/v3/merchant/4013053053
      "/path/to/apiclient_key.pem",     // 商户API证书私钥文件路径，本地文件路径
      "YOUR_PUB_KEY_ID",      // 微信支付公钥ID，如何获取请参考 https://pay.weixin.qq.com/doc/v3/merchant/4013038816
      "/path/to/wxp_pub.pem"           // 微信支付公钥文件路径，本地文件路径
    );

    CreateAbnormalRefundRequest request = new CreateAbnormalRefundRequest();
    request.refundId = "50000000382019052709732678859";
    request.outRefundNo = "1217752501201407033233368018";
    request.type = AbnormalReceiveType.MERCHANT_BANK_CARD;
    request.bankType = "ICBC_DEBIT";
    request.bankAccount = client.encrypt("bank_account");
    request.realName = client.encrypt("real_name");
    try {
      Refund response = client.run(request);
        // TODO: 请求成功，继续业务逻辑
        System.out.println(response);
    } catch (WXPayUtility.ApiException e) {
        // TODO: 请求失败，根据状态码执行不同的逻辑
        e.printStackTrace();
    }
  }

  public Refund run(CreateAbnormalRefundRequest request) {
    String uri = PATH;
    uri = uri.replace("{refund_id}", WXPayUtility.urlEncode(request.refundId));
    String reqBody = WXPayUtility.toJson(request);

    Request.Builder reqBuilder = new Request.Builder().url(HOST + uri);
    reqBuilder.addHeader("Accept", "application/json");
    reqBuilder.addHeader("Wechatpay-Serial", wechatPayPublicKeyId);
    reqBuilder.addHeader("Authorization", WXPayUtility.buildAuthorization(mchid, certificateSerialNo,privateKey, METHOD, uri, reqBody));
    reqBuilder.addHeader("Content-Type", "application/json");
    RequestBody requestBody = RequestBody.create(MediaType.parse("application/json; charset=utf-8"), reqBody);
    reqBuilder.method(METHOD, requestBody);
    Request httpRequest = reqBuilder.build();

    // 发送HTTP请求
    OkHttpClient client = new OkHttpClient.Builder().build();
    try (Response httpResponse = client.newCall(httpRequest).execute()) {
      String respBody = WXPayUtility.extractBody(httpResponse);
      if (httpResponse.code() >= 200 && httpResponse.code() < 300) {
        // 2XX 成功，验证应答签名
        WXPayUtility.validateResponse(this.wechatPayPublicKeyId, this.wechatPayPublicKey,
            httpResponse.headers(), respBody);

        // 从HTTP应答报文构建返回数据
        return WXPayUtility.fromJson(respBody, Refund.class);
      } else {
        throw new WXPayUtility.ApiException(httpResponse.code(), respBody, httpResponse.headers());
      }
    } catch (IOException e) {
      throw new UncheckedIOException("Sending request to " + uri + " failed.", e);
    }
  }

  private final String mchid;
  private final String certificateSerialNo;
  private final PrivateKey privateKey;
  private final String wechatPayPublicKeyId;
  private final PublicKey wechatPayPublicKey;

  public CreateAbnormalRefund(String mchid, String certificateSerialNo, String privateKeyFilePath, String wechatPayPublicKeyId, String wechatPayPublicKeyFilePath) {
    this.mchid = mchid;
    this.certificateSerialNo = certificateSerialNo;
    this.privateKey = WXPayUtility.loadPrivateKeyFromPath(privateKeyFilePath);
    this.wechatPayPublicKeyId = wechatPayPublicKeyId;
    this.wechatPayPublicKey = WXPayUtility.loadPublicKeyFromPath(wechatPayPublicKeyFilePath);
  }

  public String encrypt(String plainText) {
    return WXPayUtility.encrypt(this.wechatPayPublicKey, plainText);
  }

  public static class CreateAbnormalRefundRequest {
    @SerializedName("refund_id")
    @Expose(serialize = false)
    public String refundId;
  
    @SerializedName("out_refund_no")
    public String outRefundNo;
  
    @SerializedName("type")
    public AbnormalReceiveType type;
  
    @SerializedName("bank_type")
    public String bankType;
  
    @SerializedName("bank_account")
    public String bankAccount;
  
    @SerializedName("real_name")
    public String realName;
  }
  
  public static class Refund {
    @SerializedName("refund_id")
    public String refundId;
  
    @SerializedName("out_refund_no")
    public String outRefundNo;
  
    @SerializedName("transaction_id")
    public String transactionId;
  
    @SerializedName("out_trade_no")
    public String outTradeNo;
  
    @SerializedName("channel")
    public Channel channel;
  
    @SerializedName("user_received_account")
    public String userReceivedAccount;
  
    @SerializedName("success_time")
    public String successTime;
  
    @SerializedName("create_time")
    public String createTime;
  
    @SerializedName("status")
    public Status status;
  
    @SerializedName("funds_account")
    public FundsAccount fundsAccount;
  
    @SerializedName("amount")
    public Amount amount;
  
    @SerializedName("promotion_detail")
    public List<Promotion> promotionDetail;
  
  }
  
  public enum AbnormalReceiveType {
    @SerializedName("USER_BANK_CARD")
    USER_BANK_CARD,
    @SerializedName("MERCHANT_BANK_CARD")
    MERCHANT_BANK_CARD
  }
  
  public enum Channel {
    @SerializedName("ORIGINAL")
    ORIGINAL,
    @SerializedName("BALANCE")
    BALANCE,
    @SerializedName("OTHER_BALANCE")
    OTHER_BALANCE,
    @SerializedName("OTHER_BANKCARD")
    OTHER_BANKCARD
  }
  
  public enum Status {
    @SerializedName("SUCCESS")
    SUCCESS,
    @SerializedName("CLOSED")
    CLOSED,
    @SerializedName("PROCESSING")
    PROCESSING,
    @SerializedName("ABNORMAL")
    ABNORMAL
  }
  
  public enum FundsAccount {
    @SerializedName("UNSETTLED")
    UNSETTLED,
    @SerializedName("AVAILABLE")
    AVAILABLE,
    @SerializedName("UNAVAILABLE")
    UNAVAILABLE,
    @SerializedName("OPERATION")
    OPERATION,
    @SerializedName("BASIC")
    BASIC,
    @SerializedName("ECNY_BASIC")
    ECNY_BASIC
  }
  
  public static class Amount {
    @SerializedName("total")
    public Long total;
  
    @SerializedName("refund")
    public Long refund;
  
    @SerializedName("from")
    public List<FundsFromItem> from;
  
    @SerializedName("payer_total")
    public Long payerTotal;
  
    @SerializedName("payer_refund")
    public Long payerRefund;
  
    @SerializedName("settlement_refund")
    public Long settlementRefund;
  
    @SerializedName("settlement_total")
    public Long settlementTotal;
  
    @SerializedName("discount_refund")
    public Long discountRefund;
  
    @SerializedName("currency")
    public String currency;
  
    @SerializedName("refund_fee")
    public Long refundFee;
  }
  
  public static class Promotion {
    @SerializedName("promotion_id")
    public String promotionId;
  
    @SerializedName("scope")
    public PromotionScope scope;
  
    @SerializedName("type")
    public PromotionType type;
  
    @SerializedName("amount")
    public Long amount;
  
    @SerializedName("refund_amount")
    public Long refundAmount;
  
    @SerializedName("goods_detail")
    public List<GoodsDetail> goodsDetail;
  }
  
  
  public static class FundsFromItem {
    @SerializedName("account")
    public Account account;
  
    @SerializedName("amount")
    public Long amount;
  }
  
  public enum PromotionScope {
    @SerializedName("GLOBAL")
    GLOBAL,
    @SerializedName("SINGLE")
    SINGLE
  }
  
  public enum PromotionType {
    @SerializedName("CASH")
    CASH,
    @SerializedName("NOCASH")
    NOCASH
  }
  
  public static class GoodsDetail {
    @SerializedName("merchant_goods_id")
    public String merchantGoodsId;
  
    @SerializedName("wechatpay_goods_id")
    public String wechatpayGoodsId;
  
    @SerializedName("goods_name")
    public String goodsName;
  
    @SerializedName("unit_price")
    public Long unitPrice;
  
    @SerializedName("refund_amount")
    public Long refundAmount;
  
    @SerializedName("refund_quantity")
    public Long refundQuantity;
  }
  
  public enum Account {
    @SerializedName("AVAILABLE")
    AVAILABLE,
    @SerializedName("UNAVAILABLE")
    UNAVAILABLE
  }
  
}

File: skills/wechatpay-basic-payment/references/2-服务商/示例代码/Go/8-订单退款/CreateAbnormalRefund.go
package main

import (
	"bytes"
	"demo/wxpay_utility" // 引用微信支付工具库，参考 https://pay.weixin.qq.com/doc/v3/partner/4015119446
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
	"strings"
	"time"
)

func main() {
	// TODO: 请准备商户开发必要参数，参考：https://pay.weixin.qq.com/doc/v3/partner/4013080340
	config, err := wxpay_utility.CreateMchConfig(
		"YOUR_MCHID",                 // 商户号，是由微信支付系统生成并分配给每个商户的唯一标识符，商户号获取方式参考 https://pay.weixin.qq.com/doc/v3/partner/4013080340
		"YOUR_CERT_SERIAL_NO",      // 商户API证书序列号，如何获取请参考 https://pay.weixin.qq.com/doc/v3/partner/4013058924
		"/path/to/apiclient_key.pem", // 商户API证书私钥文件路径，本地文件路径
		"YOUR_PUB_KEY_ID",   // 微信支付公钥ID，如何获取请参考 https://pay.weixin.qq.com/doc/v3/partner/4013038589
		"/path/to/wxp_pub.pem",       // 微信支付公钥文件路径，本地文件路径
	)
	if err != nil {
		fmt.Println(err)
		return
	}

	request := &CreateAbnormalRefundRequest{
		RefundId:    wxpay_utility.String("50000000382019052709732678859"),
		SubMchid:    wxpay_utility.String("1900000109"),
		OutRefundNo: wxpay_utility.String("1217752501201407033233368018"),
		Type:        ABNORMALRECEIVETYPE_MERCHANT_BANK_CARD.Ptr(),
		BankType:    wxpay_utility.String("ICBC_DEBIT"),
		BankAccount: wxpay_utility.String("d+xT+MQCvrLHUVDWv/8MR/dB7TkXLVfSrUxMPZy6jWWYzpRrEEaYQE8ZRGYoeorwC+w=="), /*请传入wxpay_utility.EncryptOAEPWithPublicKey 加密结果*/
		RealName:    wxpay_utility.String("UPgQcZSdq3zOayJwZ5XLrHY2dZU1W2Cd"),                                      /*请传入wxpay_utility.EncryptOAEPWithPublicKey 加密结果*/
	}

	response, err := CreateAbnormalRefund(config, request)
	if err != nil {
		fmt.Printf("请求失败: %+v\n", err)
		// TODO: 请求失败，根据状态码执行不同的处理
		return
	}

	// TODO: 请求成功，继续业务逻辑
	fmt.Printf("请求成功: %+v\n", response)
}

func CreateAbnormalRefund(config *wxpay_utility.MchConfig, request *CreateAbnormalRefundRequest) (response *Refund, err error) {
	const (
		host   = "https://api.mch.weixin.qq.com"
		method = "POST"
		path   = "/v3/refund/domestic/refunds/{refund_id}/apply-abnormal-refund"
	)

	reqUrl, err := url.Parse(fmt.Sprintf("%s%s", host, path))
	if err != nil {
		return nil, err
	}
	reqUrl.Path = strings.Replace(reqUrl.Path, "{refund_id}", url.PathEscape(*request.RefundId), -1)
	reqBody, err := json.Marshal(request)
	if err != nil {
		return nil, err
	}
	httpRequest, err := http.NewRequest(method, reqUrl.String(), bytes.NewReader(reqBody))
	if err != nil {
		return nil, err
	}
	httpRequest.Header.Set("Accept", "application/json")
	httpRequest.Header.Set("Wechatpay-Serial", config.WechatPayPublicKeyId())
	httpRequest.Header.Set("Content-Type", "application/json")
	authorization, err := wxpay_utility.BuildAuthorization(config.MchId(), config.CertificateSerialNo(), config.PrivateKey(), method, reqUrl.RequestURI(), reqBody)
	if err != nil {
		return nil, err
	}
	httpRequest.Header.Set("Authorization", authorization)

	client := &http.Client{}
	httpResponse, err := client.Do(httpRequest)
	if err != nil {
		return nil, err
	}
	respBody, err := wxpay_utility.ExtractResponseBody(httpResponse)
	if err != nil {
		return nil, err
	}
	if httpResponse.StatusCode >= 200 && httpResponse.StatusCode < 300 {
		// 2XX 成功，验证应答签名
		err = wxpay_utility.ValidateResponse(
			config.WechatPayPublicKeyId(),
			config.WechatPayPublicKey(),
			&httpResponse.Header,
			respBody,
		)
		if err != nil {
			return nil, err
		}
		response := &Refund{}
		if err := json.Unmarshal(respBody, response); err != nil {
			return nil, err
		}

		return response, nil
	} else {
		return nil, wxpay_utility.NewApiException(
			httpResponse.StatusCode,
			httpResponse.Header,
			respBody,
		)
	}
}

type CreateAbnormalRefundRequest struct {
	RefundId    *string              `json:"refund_id,omitempty"`
	SubMchid    *string              `json:"sub_mchid,omitempty"`
	OutRefundNo *string              `json:"out_refund_no,omitempty"`
	Type        *AbnormalReceiveType `json:"type,omitempty"`
	BankType    *string              `json:"bank_type,omitempty"`
	BankAccount *string              `json:"bank_account,omitempty"`
	RealName    *string              `json:"real_name,omitempty"`
}

func (o *CreateAbnormalRefundRequest) MarshalJSON() ([]byte, error) {
	type Alias CreateAbnormalRefundRequest
	a := &struct {
		RefundId *string `json:"refund_id,omitempty"`
		*Alias
	}{
		RefundId: nil,
		Alias:    (*Alias)(o),
	}
	return json.Marshal(a)
}

type Refund struct {
	RefundId            *string        `json:"refund_id,omitempty"`
	OutRefundNo         *string        `json:"out_refund_no,omitempty"`
	TransactionId       *string        `json:"transaction_id,omitempty"`
	OutTradeNo          *string        `json:"out_trade_no,omitempty"`
	Channel             *Channel       `json:"channel,omitempty"`
	UserReceivedAccount *string        `json:"user_received_account,omitempty"`
	SuccessTime         *time.Time     `json:"success_time,omitempty"`
	CreateTime          *time.Time     `json:"create_time,omitempty"`
	Status              *Status        `json:"status,omitempty"`
	FundsAccount        *FundsAccount  `json:"funds_account,omitempty"`
	Amount              *Amount        `json:"amount,omitempty"`
	PromotionDetail     []Promotion    `json:"promotion_detail,omitempty"`
	RefundAccount       *RefundAccount `json:"refund_account,omitempty"`
}

type AbnormalReceiveType string

func (e AbnormalReceiveType) Ptr() *AbnormalReceiveType {
	return &e
}

const (
	ABNORMALRECEIVETYPE_USER_BANK_CARD     AbnormalReceiveType = "USER_BANK_CARD"
	ABNORMALRECEIVETYPE_MERCHANT_BANK_CARD AbnormalReceiveType = "MERCHANT_BANK_CARD"
)

type Channel string

func (e Channel) Ptr() *Channel {
	return &e
}

const (
	CHANNEL_ORIGINAL       Channel = "ORIGINAL"
	CHANNEL_BALANCE        Channel = "BALANCE"
	CHANNEL_OTHER_BALANCE  Channel = "OTHER_BALANCE"
	CHANNEL_OTHER_BANKCARD Channel = "OTHER_BANKCARD"
)

type Status string

func (e Status) Ptr() *Status {
	return &e
}

const (
	STATUS_SUCCESS    Status = "SUCCESS"
	STATUS_CLOSED     Status = "CLOSED"
	STATUS_PROCESSING Status = "PROCESSING"
	STATUS_ABNORMAL   Status = "ABNORMAL"
)

type FundsAccount string

func (e FundsAccount) Ptr() *FundsAccount {
	return &e
}

const (
	FUNDSACCOUNT_UNSETTLED   FundsAccount = "UNSETTLED"
	FUNDSACCOUNT_AVAILABLE   FundsAccount = "AVAILABLE"
	FUNDSACCOUNT_UNAVAILABLE FundsAccount = "UNAVAILABLE"
	FUNDSACCOUNT_OPERATION   FundsAccount = "OPERATION"
	FUNDSACCOUNT_BASIC       FundsAccount = "BASIC"
	FUNDSACCOUNT_ECNY_BASIC  FundsAccount = "ECNY_BASIC"
)

type Amount struct {
	Total            *int64          `json:"total,omitempty"`
	Refund           *int64          `json:"refund,omitempty"`
	From             []FundsFromItem `json:"from,omitempty"`
	PayerTotal       *int64          `json:"payer_total,omitempty"`
	PayerRefund      *int64          `json:"payer_refund,omitempty"`
	SettlementRefund *int64          `json:"settlement_refund,omitempty"`
	SettlementTotal  *int64          `json:"settlement_total,omitempty"`
	DiscountRefund   *int64          `json:"discount_refund,omitempty"`
	Currency         *string         `json:"currency,omitempty"`
	RefundFee        *int64          `json:"refund_fee,omitempty"`
	Advance          *int64          `json:"advance,omitempty"`
}

type Promotion struct {
	PromotionId  *string         `json:"promotion_id,omitempty"`
	Scope        *PromotionScope `json:"scope,omitempty"`
	Type         *PromotionType  `json:"type,omitempty"`
	Amount       *int64          `json:"amount,omitempty"`
	RefundAmount *int64          `json:"refund_amount,omitempty"`
	GoodsDetail  []GoodsDetail   `json:"goods_detail,omitempty"`
}

type RefundAccount string

func (e RefundAccount) Ptr() *RefundAccount {
	return &e
}

const (
	REFUNDACCOUNT_REFUND_SOURCE_PARTNER_ADVANCE      RefundAccount = "REFUND_SOURCE_PARTNER_ADVANCE"
	REFUNDACCOUNT_REFUND_SOURCE_SUB_MERCHANT         RefundAccount = "REFUND_SOURCE_SUB_MERCHANT"
	REFUNDACCOUNT_REFUND_SOURCE_SUB_MERCHANT_ADVANCE RefundAccount = "REFUND_SOURCE_SUB_MERCHANT_ADVANCE"
)

type FundsFromItem struct {
	Account *Account `json:"account,omitempty"`
	Amount  *int64   `json:"amount,omitempty"`
}

type PromotionScope string

func (e PromotionScope) Ptr() *PromotionScope {
	return &e
}

const (
	PROMOTIONSCOPE_GLOBAL PromotionScope = "GLOBAL"
	PROMOTIONSCOPE_SINGLE PromotionScope = "SINGLE"
)

type PromotionType string

func (e PromotionType) Ptr() *PromotionType {
	return &e
}

const (
	PROMOTIONTYPE_COUPON   PromotionType = "COUPON"
	PROMOTIONTYPE_DISCOUNT PromotionType = "DISCOUNT"
)

type GoodsDetail struct {
	MerchantGoodsId  *string `json:"merchant_goods_id,omitempty"`
	WechatpayGoodsId *string `json:"wechatpay_goods_id,omitempty"`
	GoodsName        *string `json:"goods_name,omitempty"`
	UnitPrice        *int64  `json:"unit_price,omitempty"`
	RefundAmount     *int64  `json:"refund_amount,omitempty"`
	RefundQuantity   *int64  `json:"refund_quantity,omitempty"`
}

type Account string

func (e Account) Ptr() *Account {
	return &e
}

const (
	ACCOUNT_AVAILABLE   Account = "AVAILABLE"
	ACCOUNT_UNAVAILABLE Account = "UNAVAILABLE"
)
